components/tfx/Evaluator/with_URI_IO/component.py

```diff
@@ -0,0 +1,103 @@
+# flake8: noqa
+
+from typing import NamedTuple
+
+def Evaluator(
+    examples_uri: 'ExamplesUri',
+    model_uri: 'ModelUri',
+    output_evaluation_uri: 'ModelEvaluationUri',
+    #output_blessing_uri: 'ModelBlessingUri',
+    baseline_model_uri: 'ModelUri' = None,
+    schema_uri: 'SchemaUri' = None,
+    eval_config: {'JsonObject': {'data_type': 'proto:tensorflow_model_analysis.EvalConfig'}} = None,
+    feature_slicing_spec: {'JsonObject': {'data_type': 'proto:tfx.components.evaluator.FeatureSlicingSpec'}} = None,
+    fairness_indicator_thresholds: list = None,
+    beam_pipeline_args: list = None,
+) -> NamedTuple('Outputs', [
+    ('evaluation_uri', 'ModelEvaluationUri'),
+    #('blessing_uri', 'ModelBlessingUri'),
+]):
+    from tfx.components import Evaluator as component_class
+
+    #Generated code
+    import json
+    import os
+    import tempfile
+    import tensorflow
+    import uuid
+    from google.protobuf import json_format, message
+    from tfx.types import channel_utils, artifact_utils
+    from tfx.components.base import base_executor
+
+    arguments = locals().copy()
+
+    component_class_args = {}
+
+    for name, execution_parameter in component_class.SPEC_CLASS.PARAMETERS.items():
+        argument_value_obj = argument_value = arguments.get(name, None)
+        if argument_value is None:
+            continue
+        parameter_type = execution_parameter.type
+        if isinstance(parameter_type, type) and issubclass(parameter_type, message.Message): # Maybe FIX: execution_parameter.type can also be a tuple
+            argument_value_obj = parameter_type()
+            json_format.Parse(argument_value, argument_value_obj)
+        component_class_args[name] = argument_value_obj
+
+    for name, channel_parameter in component_class.SPEC_CLASS.INPUTS.items():
+        artifact_path = arguments.get(name + '_uri', None) or arguments.get(name + '_path', None)
```

Comment:
This seems to rely on the assumption that the artifact argument names are suffixed by `_uri` or `_path`, following the corresponding inputs/outputs in the underlying component spec. I am wondering if there is a more manageable way of maintaining this instead of hard-coding these suffixes.

Reply:
The generated function parameters are generated in sync with this code, so they will always be in sync.
The generator adds the `_uri` or `_path` suffixes to the inputs so that the names are semantically correct. And this code corresponds to that generator behavior.
What would you like to change here?

Reply:
Just for my education, what's the difference between `_uri` and `_path`? `_uri` means a remote location like GCS while `_path` is a local artifact, is that right?

Reply:
> what's the difference between `_uri` and `_path`? `_uri` means a remote location like GCS while `_path` is a local artifact, is that right?

That's correct. `something_uri` usually has a schema/protocol and usually points to a remote location. The semantics of a URI is not well defined as it's protocol-specific. On the other hand `something_path` is a local path pointing to a file or directory. The contract for path is that it can be opened by file-related functions of any language (e.g. `open` in python) or file-based programs (e.g. `grep`).

I think that the semantic area of inputs, outputs, arguments and data passing (data vs URIs vs metadata) is pretty overloaded and it's easy to get confused. So I'm trying to use more explicit names in an effort to reduce confusion.

P.S. The '_path' suffix is a bit special for the `create_component_from_func` function. If used in conjunction with `InputPath` or `OutputPath` annotations, the input/output name will be stripped from the suffix. This is done, because the system automatically handles the transition from some abstract artifact object argument passed to the components to a file path that can be opened by the user code. So, if you have `def some_func(number_path: InputPath(int))` component function, then the component can be used as `some_op(number=42)`. It would have been weird to have `some_op(number_path=42)` and it's pretty easy to see that the type of the 42 argument ("integer") won't match the implied type of the parameter ("path"). SDK's stripping the `_path` suffix form the input name fixes the issue: `some_op(number=42)`.
The `_uri` suffix is not special, since the system does not help here and just passes URIs as opaque strings. If we had `InputUri`/`OutputUri` placeholders and annotations, then the `_uri` would have been special as well.